c2rust-analyze/src/util.rs

```diff
@@ -175,7 +185,22 @@ fn builtin_callee<'tcx>(
             Some(Callee::MiscBuiltin)
         }
 
-        _ => None,
+        "size_of" => Some(Callee::MiscBuiltin),
+
+        "malloc" => Some(Callee::Malloc),
+
+        "calloc" => Some(Callee::Calloc),
+
+        "realloc" => Some(Callee::Realloc),
+
+        "free" => Some(Callee::Free),
+
+        "is_null" => Some(Callee::IsNull),
```

Comment:
Why aren't we matching on canonical paths?

Reply:
can you explain what you mean?

Reply:
Why just `size_of` instead of `::core::mem::size_of`?  This would break if the user defined some other `size_of` function, right?

Reply:
I tested grabbing the `link_name` via `tcx.get_attr(did, rustc_span::sym::link_name)` and it's only `Some(Attribute)` if it's explicitly placed, e.g.
```rust
extern "C" {
    #[link_name = "malloc"]
    fn malloc(_: libc::c_ulong) -> *mut libc::c_void;
...
```

Reply:
Also, when checking if the `DefId` of `malloc` for example is local, `did.is_local()` returns true and `DefKind` is `Fn` and not `ForeignTy`, which seems counterintuitive for a declaration in an extern block. How else can I discover that `malloc` is an extern fn here?

Reply:
>I tested grabbing the `link_name` via `tcx.get_attr(did, rustc_span::sym::link_name)` and it's only `Some(Attribute)` if it's explicitly placed

If `link_name` is not present, then the symbol name is the name of the `fn` itself.

Also, there's a `tcx.codegen_fn_attrs` query that might be more convenient for getting the `link_name` (you don't have to parse the attribute tokens yourself that way).

>`DefKind` is `Fn` and not `ForeignTy`

`ForeignTy` is for extern types, e.g. `extern "C" { type Foo; }`.  If extern fns don't have a separate `DefKind`, you might have to check the `DefKind` of the parent item instead to see if it's an `extern` block.

Reply:
Let me know if 85f32f91 captures what's desired

Reply:
That seems fine for now.